Fix codegen icg issues

armenatzoglou opened the pull request:
In this PR, we fix several codegen issues that break some icg tests. The fixes are described below:

* In codegened slot_getattr, if there is a null attribute, then we were increasing the offset. However, in the regular version, we skip the increment for such cases. In this PR we fix this issue and increase the offset only if the processed attribute is not NULL. See story [#131305515] 
* Print if codegen is enabled/disabled at the beginning of icg/icb.

The last fix will be pushed in a separate commit. All the rest will be squashed.

@karthijrk @foyzur @hardikar Please take a look when you have some time.

hardikar commented on src/backend/codegen/slot_getattr_codegen.cc:
```diff
@@ -500,12 +508,16 @@ bool SlotGetAttrCodegen::GenerateSlotGetAttr(
+    irb->CreateStore(irb->CreateCall(llvm_att_align_nominal, {
+        irb->CreateLoad(llvm_off_ptr),
+        codegen_utils->GetConstant<char>(thisatt->attalign)}),
+                     llvm_off_ptr);
```
Can you re-align this statement, to cleanly show the first parameter and second paramter to CreateStore.

armenatzoglou replied:
Will do it.

hardikar commented on src/backend/codegen/slot_getattr_codegen.cc:
```diff
@@ -500,12 +508,16 @@ bool SlotGetAttrCodegen::GenerateSlotGetAttr(
 
-    off = att_align_nominal(off, thisatt->attalign);
+    // off = att_align_nominal(off, thisatt->attalign);
+    irb->CreateStore(irb->CreateCall(llvm_att_align_nominal, {
```
Another downside of using the external function is that the (LLVM) compiler can no longer optimize the constants for the case when all the attributes are NON-NULL. If you only did { x = 1; x=x+1; x=x+1 }, it could determine the constants at each stage, with an external call in the middle { x = 1; x=f(x); x=x+1}, that is out.

hardikar commented on src/test/regress/pg_regress.c:
```diff
@@ -2478,7 +2492,14 @@ regression_main(int argc, char *argv[], init_function ifunc, test_function tfunc
 	 */
-	check_optimizer_status();
+	check_feature_status("optimizer",
+						 "Optimizer enabled. Using optimizer answer files whenever possible",
```
Indentation.

hardikar commented on src/test/regress/pg_regress.c:
```diff
@@ -2073,15 +2080,22 @@ check_optimizer_status(void)
 		{
-			optimizer_enabled = true;
-			status(_("Optimizer enabled. Using optimizer answer files whenever possible"));
+			if (strncmp(feature_name, "optimizer", 9) == 0)
```
The `check_feature_status` abstraction is half baked IMHO. You have an abstracted function that has a special case for one of it's inputs. This should probably be implemented on of these ways
* a separate function- that can be a wrapper for the special case of optimizer, or a callback functions that's called doing the appropriate action
* take a pointer in this function and set that accordingly.

armenatzoglou replied:
This is a really good catch. Thanks a lot! I will fix it.

armenatzoglou commented on src/backend/codegen/slot_getattr_codegen.cc:
```diff
@@ -363,7 +366,12 @@ bool SlotGetAttrCodegen::GenerateSlotGetAttr(
 
-  int off = 0;
+  // int off = 0;
+  llvm::Value* llvm_off_ptr = irb->CreateAlloca(
```
Good point, I will change it.

armenatzoglou commented:
@hardikar We have addressed your comments. Please take a look. Also, we have reverted the fix that avoids to error out when `advance_aggregates` processes a null attribute because we do not fall back properly.

armenatzoglou commented:
Merged: 96f22dac357e04b6e950ca7d36a852b45cb08d5a and 12cfd7bd5493fcd709c508f625897b7f540b5877. 

Thanks a lot for the review!

Nikos & Nikhil
